File: day_10/rust/src/main.rs
use std::cell::RefCell;
use std::fmt::{Display, Formatter};
use std::fs;
use std::rc::Rc;

type TilePtr = Rc<RefCell<MazeTile>>;

#[derive(Debug, Eq, PartialEq)]
enum Direction {
    North,
    South,
    East,
    West,
}

#[derive(Debug, Eq, PartialEq)]
struct Connection {
    direction: Direction,
    tile: TilePtr,
}

#[derive(Debug, Default, Eq, PartialEq)]
struct Pipe {
    connection_0: Option<Connection>,
    connection_1: Option<Connection>,
    is_start: bool,
}

#[derive(Debug, PartialEq, Eq)]
enum MazeTile {
    Vertical(Pipe),
    Horizontal(Pipe),
    NorthEastBend(Pipe),
    NorthWestBend(Pipe),
    SouthWestBend(Pipe),
    SouthEastBend(Pipe),
    Ground,
    Start,
}

#[derive(Debug, PartialEq, Eq)]
struct ParseMazeTileError;

impl TryFrom<char> for MazeTile {
    type Error = ParseMazeTileError;

    fn try_from(value: char) -> Result<Self, Self::Error> {
        match value {
            '|' => Ok(Self::Vertical(Pipe::default())),
            '-' => Ok(Self::Horizontal(Pipe::default())),
            'L' => Ok(Self::NorthEastBend(Pipe::default())),
            'J' => Ok(Self::NorthWestBend(Pipe::default())),
            '7' => Ok(Self::SouthWestBend(Pipe::default())),
            'F' => Ok(Self::SouthEastBend(Pipe::default())),
            '.' => Ok(Self::Ground),
            'S' => Ok(Self::Start),
            _ => Err(ParseMazeTileError),
        }
    }
}

impl Display for MazeTile {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        let symbol = match self {
            MazeTile::Vertical(_) => '┃',
            MazeTile::Horizontal(_) => '━',
            MazeTile::NorthEastBend(_) => '┗',
            MazeTile::NorthWestBend(_) => '┛',
            MazeTile::SouthWestBend(_) => '┓',
            MazeTile::SouthEastBend(_) => '┏',
            MazeTile::Ground => '░',
            MazeTile::Start => 'S',
        };

        write!(f, "{symbol}")
    }
}

fn connect_pipes(tiles: &mut [[TilePtr]]) {
    for (i, row) in tiles.iter_mut().enumerate() {}
}

const TEST_INPUT: &str = ".....\n\
.F-7.\n\
.|.|.\n\
.L-J.\n\
.....";

fn main() {
    let tiles: Vec<Vec<TilePtr>> = fs::read_to_string("input.txt")
        .expect("failed to open input file")
        .lines()
        .map(|line| {
            line.chars()
                .filter_map(|c| c.try_into().ok())
                .map(|tile| Rc::new(RefCell::new(tile)))
                .collect()
        })
        .collect();

    let mut out_str = String::new();
    for row in tiles {
        for tile in row {
            out_str.push_str(&*format!("{tile}"));
        }
        out_str.push('\n');
    }

    print!("{out_str}");
}

File: day_10/ts/src/main.ts
import { readFile } from "fs/promises";

enum TileType {
    VERTICAL = '┃',
    HORIZONTAL = '━',
    NORTH_EAST_BEND = '┗',
    NORTH_WEST_BEND = '┛',
    SOUTH_WEST_BEND = '┓',
    SOUTH_EAST_BEND = '┏',
    GROUND = '▒',
    START = '╳'
}

class Position {
    public x: number;
    public y: number;

    constructor(x: number = 0, y: number = 0) {
        this.x = x;
        this.y = y;
    }

    public eq(other: Position): boolean {
        return (this.x === other.x) && (this.y === other.y);
    }

    public neq(other: Position): boolean {
        return !this.eq(other);
    }
}



class MazeTile {
    static readonly CHAR_TILE_MAP = new Map([
        ['|', TileType.VERTICAL],
        ['-', TileType.HORIZONTAL],
        ['L', TileType.NORTH_EAST_BEND],
        ['J', TileType.NORTH_WEST_BEND],
        ['7', TileType.SOUTH_WEST_BEND],
        ['F', TileType.SOUTH_EAST_BEND],
        ['.', TileType.GROUND],
        ['S', TileType.START]
    ]);

    static readonly PRINT_CHAR_MAP = new Map([
        [TileType.VERTICAL, '┃'],
        [TileType.HORIZONTAL, '━'],
        [TileType.NORTH_EAST_BEND, '┗'],
        [TileType.NORTH_WEST_BEND, '┛'],
        [TileType.SOUTH_WEST_BEND, '┓'],
        [TileType.SOUTH_EAST_BEND, '┏'],
        [TileType.GROUND, '░'],
        [TileType.START, '╳']
    ]);

    private type: TileType;
    private isStart: boolean;
    private connectedToStart: boolean;
    private position: Position;

    /**
     * Construct a new MazeTile instance from a given character
     * @param c a string containing a single character
     */
    constructor(c: string, x: number, y: number) {
        this.type = MazeTile.CHAR_TILE_MAP.get(c) ?? TileType.GROUND;
        this.isStart = this.type == TileType.START;
        this.connectedToStart = this.isStart;
        this.position = new Position(x, y);
    }

    public toString(): string {
        const bkgColor: number = this.isStart ? 44 : 40;
        const foregroundColor: number = this.connectedToStart ? 33 : 37;
        const tileColor = `\x1b[${foregroundColor};${bkgColor}m`;

        const tileChar = MazeTile.PRINT_CHAR_MAP.get(this.type) ?? '?';

        return `${tileColor}${tileChar}\x1b[0`;
    }

    public static printTiles(tiles: Array<Array<MazeTile>>) {
        let outStr = "";
        for (const row of tiles) {
            for (const tile of row) {
                outStr += tile.toString();
            }
            outStr += '\n';
        }
        console.log(outStr);
    }

    private determinePipeType(tiles: Array<Array<MazeTile>>) {
        if (this.type != TileType.START) {
            return;
        }
        const pos = this.position;

        const prevRow = pos.y - 1;
        const nextRow = pos.y + 1;
        const prevCol = pos.x - 1;
        const nextCol = pos.x + 1;

        const northPipeType = prevRow >= 0 ? tiles[prevRow][pos.x].type : null;
        const southPipeType = nextRow < tiles.length ? tiles[nextRow][pos.x].type : null;
        const westPipeType = prevCol >= 0 ? tiles[pos.y][prevCol].type : null;
        const eastPipeType = nextCol < tiles[pos.y].length ? tiles[pos.y][nextCol].type : null;

        const connectsNorth: boolean = northPipeType != null &&
            [TileType.VERTICAL, TileType.SOUTH_WEST_BEND, TileType.SOUTH_EAST_BEND].includes(northPipeType);
        const connectsSouth: boolean = southPipeType != null &&
            [TileType.VERTICAL, TileType.NORTH_WEST_BEND, TileType.NORTH_EAST_BEND].includes(southPipeType);
        const connectsEast: boolean = eastPipeType != null &&
            [TileType.HORIZONTAL, TileType.SOUTH_WEST_BEND, TileType.NORTH_WEST_BEND].includes(eastPipeType);
        const connectsWest: boolean = westPipeType != null &&
            [TileType.HORIZONTAL, TileType.SOUTH_EAST_BEND, TileType.NORTH_EAST_BEND].includes(westPipeType);

        let startType = TileType.START;
        if (connectsNorth && connectsSouth) {
            startType = TileType.VERTICAL;
        } else if (connectsEast && connectsWest) {
            startType = TileType.HORIZONTAL;
        } else if (connectsNorth && connectsEast) {
            startType = TileType.NORTH_EAST_BEND;
        } else if (connectsNorth && connectsWest) {
            startType = TileType.NORTH_WEST_BEND;
        } else if (connectsSouth && connectsEast) {
            startType = TileType.SOUTH_EAST_BEND;
        } else if (connectsSouth && connectsWest) {
            startType = TileType.SOUTH_WEST_BEND;
        }

        this.type = startType;
    }

    /**
     * Locates starting tile, determines it's type, and returns its position
     * @param tiles 
     * @returns starting tile's position
     */
    public static setupStartTile(tiles: Array<Array<MazeTile>>): MazeTile | null {
        let tile = null;
        for (let y = 0; y < tiles.length; y++) {
            for (let x = 0; x < tiles[y].length; x++) {
                const tmp = tiles[y][x];
                if (tmp.type == TileType.START) {
                    tile = tmp;
                }
            }
        }

        if (tile) {
            // Determine type of start pipe
            tile.determinePipeType(tiles);
        }

        return tile;
    }

    private getConnections(): { con0: Position, con1: Position } {
        const type = this.type;
        const pos = this.position;
        let con0 = new Position();
        let con1 = new Position();
        switch (type) {
            case TileType.HORIZONTAL:
                con0 = new Position(pos.x - 1, pos.y);
                con1 = new Position(pos.x + 1, pos.y);
                break;
            case TileType.VERTICAL:
                con0 = new Position(pos.x, pos.y - 1);
                con1 = new Position(pos.x, pos.y + 1);
                break;
            case TileType.NORTH_EAST_BEND:
                con0 = new Position(pos.x, pos.y - 1);
                con1 = new Position(pos.x + 1, pos.y);
                break;
            case TileType.NORTH_WEST_BEND:
                con0 = new Position(pos.x, pos.y - 1);
                con1 = new Position(pos.x - 1, pos.y);
                break;
            case TileType.SOUTH_EAST_BEND:
                con0 = new Position(pos.x, pos.y + 1);
                con1 = new Position(pos.x + 1, pos.y);
                break;
            case TileType.SOUTH_WEST_BEND:
                con0 = new Position(pos.x, pos.y + 1);
                con1 = new Position(pos.x - 1, pos.y);
                break;
        }

        return { con0, con1 };
    }

    public static traversePipes(tiles: Array<Array<MazeTile>>, startTile: MazeTile): number {
        let connections = startTile.getConnections();

        let prev = startTile;
        let next = tiles[connections.con0.y][connections.con0.x];
        next.connectedToStart = true;

        let stepCount = 0;
        while (next.position.neq(startTile.position)) {
            stepCount++;

            connections = next.getConnections();
            let tmpTile = tiles[connections.con0.y][connections.con0.x];
            if (tmpTile === prev) {
                tmpTile = tiles[connections.con1.y][connections.con1.x];
            }
            prev = next;
            next = tmpTile;
            next.connectedToStart = true;
        }
        return Math.ceil(stepCount / 2);
    }
}

const TEST_INPUT: string = `7-F7-
.FJ|7
SJLL7
|F--J
LJ.LJ
`;

async function main() {
    const input_str = await readFile("input.txt", { encoding: 'utf-8' });
    const tiles: Array<Array<MazeTile>> = [];
    const lines = input_str.split(/\n\r?/);
    for (let y = 0; y < lines.length; y++) {
        const row: MazeTile[] = [];
        const chars = lines[y].split('');
        for (let x = 0; x < chars.length; x++) {
            const c = chars[x];
            row.push(new MazeTile(c, x, y));
        }
        tiles.push(row);
    }

    const startTile = MazeTile.setupStartTile(tiles);
    if (!startTile) {
        throw new Error("Failed to find staring pipe");
    }

    const furthestPt = MazeTile.traversePipes(tiles, startTile);
    MazeTile.printTiles(tiles);
    console.log(`Part 1 result: ${furthestPt}`);
}

main()
    .catch((reason) => {
        console.error(reason);
    });
